base/src/H264DecoderV4L2Helper.cpp

```diff
@@ -0,0 +1,1478 @@
+        }
+    }
+ 
+    /* Start stream processing on output plane
+    ** by setting the streaming status ON.
+    */
+ 
+    ret = v4l2_ioctl(ctx.fd,VIDIOC_STREAMON, &ctx.op_buf_type);
+    if (ret != 0)
+    {
+        LOG_ERROR << "Streaming error on output plane" << endl;
+        ctx.in_error = 1;
+    }
+ 
+    ctx.op_streamon = 1;
+    // Create Capture loop thread.
+    typedef void * (*THREADFUNCPTR)(void *);
+   
+    pthread_create(&ctx.dec_capture_thread, NULL,h264DecoderV4L2Helper::capture_thread, (void *) (this));
+}
+int h264DecoderV4L2Helper::process(frame_sp inputFrame)
+{
+    uint32_t idx = 0;
+    while (!ctx.eos && !ctx.in_error && idx < ctx.op_num_buffers)
+    {
+        struct v4l2_buffer queue_v4l2_buf_op;
+        struct v4l2_plane queue_op_planes[MAX_PLANES];
+        Buffer *buffer;
+ 
+        memset(&queue_v4l2_buf_op, 0, sizeof (queue_v4l2_buf_op));
+        memset(queue_op_planes, 0, sizeof (queue_op_planes));
+ 
+        buffer = ctx.op_buffers[idx];
+        if (ctx.decode_pixfmt == V4L2_PIX_FMT_H264)
+        {
+            read_input_chunk_frame_sp(inputFrame, buffer);
+        }
+        else
+        {
+            LOG_ERROR << "Currently only H264 supported" << endl;
+            ctx.in_error = 1;
+        }
+ 
+        queue_v4l2_buf_op.index = idx;
+        queue_v4l2_buf_op.m.planes = queue_op_planes;
+ 
+        /* Enqueue the buffer on output plane
+        ** It is necessary to queue an empty buffer
+        ** to signal EOS to the decoder.
+        */
+        ret = q_buffer(&ctx, queue_v4l2_buf_op, buffer,
+            ctx.op_buf_type, ctx.op_mem_type, ctx.op_num_planes);
+        if (ret)
+        {
+            LOG_ERROR << "Error Qing buffer at output plane" << endl;
+            ctx.in_error = 1;
+        }
+ 
+        if (queue_v4l2_buf_op.m.planes[0].bytesused == 0)
+        {
+            ctx.eos = true;
+            LOG_INFO << "Input file read complete" << endl;
+            break;
+        }
+        idx++;
+    }
+ 
+    // Dequeue and queue loop on output plane.
+    ctx.eos = decode_process(ctx,inputFrame);
+   
+    /* For blocking mode, after getting EOS on output plane,
+    ** dequeue all the queued buffers on output plane.
+    ** After that capture plane loop should be signalled to stop.
+    *///
+ 
+    while (ctx.num_queued_op_buffers > 0 && !ctx.in_error && !ctx.got_eos)
+    {
+        struct v4l2_buffer v4l2_buf;
+        struct v4l2_plane planes[MAX_PLANES];
+ 
+        memset(&v4l2_buf, 0, sizeof (v4l2_buf));
+        memset(planes, 0, sizeof (planes));
+ 
+        v4l2_buf.m.planes = planes;
+        ret = dq_buffer(&ctx, v4l2_buf, NULL, ctx.op_buf_type, ctx.op_mem_type, -1);
+        if (ret)
+        {
+            LOG_ERROR << "Error DQing buffer at output plane" << endl;
+            ctx.in_error = 1;
+            break;
+            return true;
+        }
+    }
+ 
+    // Signal EOS to capture loop thread to exit.
+ 
+    if(ctx.eos == true)
```

Comment:
We are handling EOS in the decode_process() function where we have a check if EOS is true and deque the empty buffer on the output plane.

Reply:
why 2 flags ? @mohammedzakikochargi

Reply:
2 flags are almost the same . There are 2 while loops in the process  ,for the EOS frame i make ctx.eos as true in the first while loop because ctx.eos is used as the check in some different function after while loop. wherein i still need ctx.got_eos as false for the second while loop , and i make ctx.got_eos as true after 2nd while loop is done for the EOS frame.